### ISSUE TITLE
Unused `tuple`, `dict`, `set` and `list`  literals are not reported per `reportUnusedExpression`

### ISSUE THREAD

**Issue by Bobronium** [microsoft/pyright#3896]
**Describe the bug**
Unused `tuple`, `dict`, `set` and `list`  literals are not reported per`reportUnusedExpression`

**To Reproduce**
1. Set `reportUnusedExpression` to `warning`, `error` or `information`
2.  Typecheck the code from example

**Expected behavior**
Pyright reports `Expression value is unused` for each line of the file

**Screenshots or Code**
<img width="200" alt="image" src="https://user-images.githubusercontent.com/36469655/188032820-f63ecfab-eda9-4115-bc49-2c037e64adb6.png">
<img width="565" alt="image" src="https://user-images.githubusercontent.com/36469655/188032861-d4eb6efa-5686-40ac-bc4f-14f923e15111.png">

```python
object
1
None
True
()
{}
{...}
[]
(), {}, {...}, []
() or ()
```

**Additional context**
Pylance Language Server version: 2022.8.50
OS and version: darwin arm64
Python version (& distribution if applicable, e.g. Anaconda): 3.10.0, pyenv
python.analysis.indexing: true
python.analysis.typeCheckingMode: basic

**Comment by erictraut** [comment]
This is by design. The primary purpose of this diagnostic rule is to catch cases where an equals comparison is used in place of an assignment statement.

See https://github.com/microsoft/pyright/issues/3174 for details.

I'm willing to entertain an expansion of this check if there's a good justification for it. Are you asserting that an unused tuple, dict, set and list expression is likely to be unintentional and the source of common bugs?

**Comment by Bobronium** [comment]
> This is by design. The primary purpose of this diagnostic rule is to catch cases where an equals comparison is used in place of an assignment statement.

But none of unused expressions from my example have anything to do with equals operator, and yet all of theme are reported except collections.

I believe justification for broadening the scope of the rule can be found in the exact issue you've linked.

The question is: what makes collections literals different from other literals and identifiers?

Why `1` is reported and `{1}` is not, despite both being unused expressions?

**Comment by erictraut** [comment]
I implemented the change, but I backed out most of it because it generated too many false positives in internal Microsoft code bases. The problem is that there are legitimate uses of comprehensions within list, set, and dictionary expressions. Such comprehensions are sometimes used as a concise looping construct even if the resulting list, set, or dictionary are never used.

I did extend the check to report unused tuple expressions because I can't think of any legitimate uses for these. This caught several (harmless) bugs in the code bases I tested it on.

**Comment by Bobronium** [comment]
Can pyright differentiate between comprehensions and regular constructions and bark only on the latter?

I also would argue that using comprehensions for quick looping is rather an anti-pattern and may not obvious. 

One option would be to limit this to expressions that don't have potential side effects.

I think PyCharm makes a great job differentiating between definitely unused expressions and ones that may have potential side effects.

**Comment by erictraut** [comment]
I'm going to leave it as is for now. I might reconsider in the future if there's sufficient signal from other pyright users.

**Comment by erictraut** [comment]
On second thought, I tried implementing your suggestion of filtering expressions with comprehensions. That produced no false positives on any of our internal code bases, so I think it's safe to report unused dict, list and set expressions as long as they don't include comprehensions. This will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
